File: pulsecheck/lib/services/health_repository.dart
import '../db/app_database.dart';
import '../models/blood_pressure_reading.dart';
import '../models/heart_rate_reading.dart';

/// All reads/writes to the local health database go through here — the
/// rest of the app never touches SQL directly.
class HealthRepository {
  final AppDatabase _db = AppDatabase.instance;

  Future<int> addHeartRate(HeartRateReading reading) async {
    final db = await _db.database;
    return db.insert('heart_rate_readings', reading.toMap());
  }

  Future<int> addBloodPressure(BloodPressureReading reading) async {
    final db = await _db.database;
    return db.insert('blood_pressure_readings', reading.toMap());
  }

  Future<void> deleteHeartRate(int id) async {
    final db = await _db.database;
    await db.delete('heart_rate_readings', where: 'id = ?', whereArgs: [id]);
  }

  Future<void> deleteBloodPressure(int id) async {
    final db = await _db.database;
    await db.delete('blood_pressure_readings', where: 'id = ?', whereArgs: [id]);
  }

  Future<List<HeartRateReading>> heartRateReadings({DateTime? since}) async {
    final db = await _db.database;
    final rows = await db.query(
      'heart_rate_readings',
      where: since != null ? 'taken_at >= ?' : null,
      whereArgs: since != null ? [since.toIso8601String()] : null,
      orderBy: 'taken_at DESC',
    );
    return rows.map(HeartRateReading.fromMap).toList();
  }

  Future<List<BloodPressureReading>> bloodPressureReadings({DateTime? since}) async {
    final db = await _db.database;
    final rows = await db.query(
      'blood_pressure_readings',
      where: since != null ? 'taken_at >= ?' : null,
      whereArgs: since != null ? [since.toIso8601String()] : null,
      orderBy: 'taken_at DESC',
    );
    return rows.map(BloodPressureReading.fromMap).toList();
  }

  Future<HeartRateReading?> latestHeartRate() async {
    final rows = await heartRateReadings();
    return rows.isEmpty ? null : rows.first;
  }

  Future<BloodPressureReading?> latestBloodPressure() async {
    final rows = await bloodPressureReadings();
    return rows.isEmpty ? null : rows.first;
  }

  Future<void> clearAll() async {
    final db = await _db.database;
    await db.delete('heart_rate_readings');
    await db.delete('blood_pressure_readings');
  }
}

File: vitaltrack/Sources/VitalTrackCore/Persistence/HealthRepository.swift
import CoreData
import Foundation

/// Everything the app reads/writes for heart rate and blood pressure goes
/// through this protocol — view models never touch Core Data directly,
/// which is also what makes them testable with an in-memory fake.
public protocol HealthRepository: Sendable {
    func addHeartRate(_ reading: HeartRateReading) async throws
    func addBloodPressure(_ reading: BloodPressureReading) async throws

    func deleteHeartRate(id: UUID) async throws
    func deleteBloodPressure(id: UUID) async throws

    func heartRateReadings(since: Date?) async throws -> [HeartRateReading]
    func bloodPressureReadings(since: Date?) async throws -> [BloodPressureReading]

    func latestHeartRate() async throws -> HeartRateReading?
    func latestBloodPressure() async throws -> BloodPressureReading?

    func deleteAll() async throws
}

public final class CoreDataHealthRepository: HealthRepository, @unchecked Sendable {
    private let persistence: PersistenceController

    public init(persistence: PersistenceController = .shared) {
        self.persistence = persistence
    }

    public func addHeartRate(_ reading: HeartRateReading) async throws {
        let context = persistence.newBackgroundContext()
        try await context.perform {
            let entity = HeartRateReadingEntity(context: context)
            entity.apply(from: reading)
            try self.persistence.saveIfNeeded(context)
        }
    }

    public func addBloodPressure(_ reading: BloodPressureReading) async throws {
        let context = persistence.newBackgroundContext()
        try await context.perform {
            let entity = BloodPressureReadingEntity(context: context)
            entity.apply(from: reading)
            try self.persistence.saveIfNeeded(context)
        }
    }

    public func deleteHeartRate(id: UUID) async throws {
        let context = persistence.newBackgroundContext()
        try await context.perform {
            let request = HeartRateReadingEntity.fetchRequest()
            request.predicate = NSPredicate(format: "id == %@", id as CVarArg)
            for object in try context.fetch(request) { context.delete(object) }
            try self.persistence.saveIfNeeded(context)
        }
    }

    public func deleteBloodPressure(id: UUID) async throws {
        let context = persistence.newBackgroundContext()
        try await context.perform {
            let request = BloodPressureReadingEntity.fetchRequest()
            request.predicate = NSPredicate(format: "id == %@", id as CVarArg)
            for object in try context.fetch(request) { context.delete(object) }
            try self.persistence.saveIfNeeded(context)
        }
    }

    public func heartRateReadings(since: Date? = nil) async throws -> [HeartRateReading] {
        let context = persistence.newBackgroundContext()
        return try await context.perform {
            let request = HeartRateReadingEntity.fetchRequest()
            if let since {
                request.predicate = NSPredicate(format: "takenAt >= %@", since as NSDate)
            }
            request.sortDescriptors = [NSSortDescriptor(key: "takenAt", ascending: false)]
            return try context.fetch(request).map { $0.toModel() }
        }
    }

    public func bloodPressureReadings(since: Date? = nil) async throws -> [BloodPressureReading] {
        let context = persistence.newBackgroundContext()
        return try await context.perform {
            let request = BloodPressureReadingEntity.fetchRequest()
            if let since {
                request.predicate = NSPredicate(format: "takenAt >= %@", since as NSDate)
            }
            request.sortDescriptors = [NSSortDescriptor(key: "takenAt", ascending: false)]
            return try context.fetch(request).map { $0.toModel() }
        }
    }

    public func latestHeartRate() async throws -> HeartRateReading? {
        try await heartRateReadings(since: nil).first
    }

    public func latestBloodPressure() async throws -> BloodPressureReading? {
        try await bloodPressureReadings(since: nil).first
    }

    public func deleteAll() async throws {
        let context = persistence.newBackgroundContext()
        try await context.perform {
            for entityName in ["HeartRateReadingEntity", "BloodPressureReadingEntity"] {
                let request = NSFetchRequest<NSFetchRequestResult>(entityName: entityName)
                let delete = NSBatchDeleteRequest(fetchRequest: request)
                try context.execute(delete)
            }
            try self.persistence.saveIfNeeded(context)
        }
    }
}
